### PR TITLE
Update GAS distribution logic, part1

### PR DESCRIPTION
**Pull request by fyrchik**
Related #1348 .
Implement distribution to NEO holders and validators.

### CODE REVIEW

File: pkg/core/native/native_neo.go [hunk 1]
```diff
@@ -183,10 +183,22 @@ func (n *NEO) Initialize(ic *interop.Context) error {
...
+	}
+	index := int(ic.Block.Index) % len(ic.Chain.GetConfig().StandbyCommittee)
+	gas.Mul(gas, big.NewInt(committeeRewardRatio))
+	n.GAS.mint(ic, pubs[index].GetScriptHash(), gas.Div(gas, big.NewInt(100)))
+
 	if !n.votesChanged.Load().(bool) {
 		return nil
 	}
-	pubs, err := n.GetValidatorsInternal(ic.Chain, ic.DAO)
+	pubs, err = n.GetValidatorsInternal(ic.Chain, ic.DAO)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:201]: It'd be nice to split `GetValidatorsInternal` to not call `GetCommitteeMembers()` twice (it has some cost to it).

File: pkg/core/native/native_neo.go [hunk 2]
```diff
@@ -183,10 +183,22 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 
 // OnPersist implements Contract interface.
 func (n *NEO) OnPersist(ic *interop.Context) error {
+	gas, err := n.GetGASPerBlock(ic, ic.Block.Index)
+	if err != nil {
+		return err
+	}
+	pubs, err := n.GetCommitteeMembers(ic.Chain, ic.DAO)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:190]: We've just optimized out `GetValidatorsInternal` in #1337, maybe we can do the same with this? We can't avoid DB write this time, unfortunately, but at least we can avoid recalculating the committee.

File: pkg/core/native/native_neo.go [hunk 3]
```diff
@@ -53,6 +53,18 @@ const (
 	effectiveVoterTurnout = 5
 )
 
+const (
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go]: Fixed.

File: pkg/core/native/native_neo.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:186]: Fixed.

File: pkg/core/native/native_neo.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:197]: Fixed (`GetCommitteeMembers` is not cached)

**Comment by roman-khimov** [comment]
Linter still isn't happy.

File: pkg/core/native/native_neo.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/native/native_neo.go:197]: But I don't see any change wrt this issue.

File: pkg/core/native/native_neo.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/core/native/native_neo.go:186]: And this one too.

File: pkg/core/native/native_neo.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:199]: Added `committee` caching.

**Review by roman-khimov — COMMENTED** [review]
Looks good now, but I'll let it hang for a while until we finish our TPS improvements (this one may affect it).

**Comment by roman-khimov** [comment]
Please, rebase it. It's outdated a little wrt C# implementation, but it's a nice starting point.

**Comment by roman-khimov** [comment]
I don't see changes from neo-project/neo#1884 there (like PostPersist). Same thing with neo-project/neo#1917.

**Comment by fyrchik** [comment]
@roman-khimov changes in neo-project/neo#1917 are overriden by ones from neo-project/neo#1920 .

I wanted to do `postPersist` later, though nothing stops from doing it now.

**Review by roman-khimov — APPROVED** [review]
We still have duplicate AER problem, but let's create a separate issue for that.

**Pull request merged**